kratos/mpi/utilities/mpi_search_utilities.h

```diff
@@ -0,0 +1,284 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+#pragma once
+
+// System includes
+#include <vector>
+#include <numeric>
+
+// External includes
+
+// Project includes
+#include "includes/data_communicator.h"
+
+namespace Kratos
+{
+
+/**
+ * @class MPISearchUtilities
+ * @ingroup KratosCore
+ * @brief This class provides the utilities for MPI search
+ * @author Vicente Mataix Ferrandiz
+ */
+class MPISearchUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+        
+    /// Define zero tolerance as Epsilon
+    static constexpr double ZeroTolerance = std::numeric_limits<double>::epsilon();
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief MPISynchronousPointSynchronization prepares synchronously the coordinates of the points for MPI search.
+     * @param itPointBegin Iterator to the beginning of the points range
+     * @param itPointEnd Iterator to the end of the points range
+     * @param rAllPointsCoordinates vector where the computed coordinates will be stored
+     * @param rDataCommunicator The data communicator
+     * @param AllPointsAreTheSame If all points are the same (-1 to be computed, 0 if not, 1 if yes)
+     * @param NumberOfPoints The number of points in the range
+     * @param TotalNumberOfPoints The total number of points in all the ranges
+     * @return The number of points in the range
+     * @tparam TPointIteratorType The type of the point iterator
+     */
+    template<typename TPointIteratorType>
+    static int MPISynchronousPointSynchronization(
+        TPointIteratorType itPointBegin,
+        TPointIteratorType itPointEnd,
+        std::vector<double>& rAllPointsCoordinates,
+        const DataCommunicator& rDataCommunicator,
+        int AllPointsAreTheSame = -1,
+        int NumberOfPoints = -1,
+        int TotalNumberOfPoints = -1
+        )
+    {
+        // First check that the points are the same in all processes
+        if (AllPointsAreTheSame < 0) {
+            AllPointsAreTheSame = static_cast<int>(CheckAllPointsAreTheSame(itPointBegin, itPointEnd, NumberOfPoints, TotalNumberOfPoints, rDataCommunicator));
```

Comment:
There are several things here that I don't fully understand. I put is as a comment of this function because they are mainly related with this.

1 - If function does anything else than checking if all the points are the same, for example calculating `TotalNumberOfElements`, it should not be called check. Otherwise is very misleading.

2 - `TotalNumberOfPoints` is the actual total or the different ones? If is the total, it would be more performant to just calculate that number outside the function in order to prevent the the collective in line 97 (the `AllGather) which essentially you are using for the same (`SumAll` = `AllGather` + `Reduce`, but you are repeating the `AllGather` below). If the total number is the different ones, then either `Check` does something more behind the scenes, or the line 97 collective is calculating something incorrect.

Reply:
1- Okay, improving name.
2- TotalNumberOfPoints is the sum of the nodes in all partitions, I will improve name.